lib/fs/promises.js

```diff
@@ -147,15 +147,15 @@ async function readFileHandle(filehandle, options) {
 
   const chunks = [];
   const chunkSize = Math.min(size, 16384);
-  const buf = Buffer.alloc(chunkSize);
   let totalRead = 0;
   do {
+    const buf = Buffer.alloc(chunkSize);
     const { bytesRead, buffer } =
-      await read(filehandle, buf, 0, buf.length);
-    totalRead = bytesRead;
-    if (totalRead > 0)
-      chunks.push(buffer.slice(0, totalRead));
-  } while (totalRead === chunkSize);
+      await read(filehandle, buf, 0, chunkSize, totalRead);
+    totalRead += bytesRead;
+    if (bytesRead > 0)
+      chunks.push(buffer.slice(0, bytesRead));
+  } while (totalRead < size);
```

Comment:
`size` cannot be used as a reliable indicator of when the loop should exit since the file size might change while `readFileHandle()` is running. Maybe what we could do is using `bytesRead === chunkSize` as the exit condition.